### PR TITLE
Milestone 3: OOP school library: set up associations

### PR DESCRIPTION
**Pull request by prg-04**
### I have done the following to reach this milestone:

- [ ] Created a class Classroom with the following:
     - [ ] @label instance variable, should be initialized in the constructor.
     - [ ] Setter and getter for @label (remember about attr_accessor).

- [ ] Created the has-many/belongs-to relationship between Classroom and Student. The following was implemented:
     - [ ] Created the has-many side (a classroom has many students).
     - [ ] Created the belongs-to side (a student belongs to a classroom).
     - [ ] Made sure that when adding a student to a classroom it also sets the classroom for the student.
     - [ ] Made sure that when setting the classroom for a student it also adds it to the classrooms' students.

- [ ] Created a class Book with the following:
     - [ ] @title and @author instance variables, should be initialized in the constructor.
     - [ ] Setters and getters for instance variables (remember about attr_accessor).

- [ ] Create a class Rental with the following:
     - [ ] @date instance variable, should be initialized in the constructor.
     - [ ] Setter and getter for @date (remember about attr_accessor).
- [ ] Created the many-to-many (also has-many-through) relationship between Person and Book using the intermediate class Rental. The following was implemented:
     - [ ] Created the has-many side of Book and Rental (a book has many rentals).
     - [ ] Created the belongs-to side of Rental and Book (a rental belongs to a book).
     - [ ] Created the has-many side of Person and Rental (a person has many rentals).
     - [ ] Created the belongs-to side of Rental and Person (a rental belongs to a person).
     - [ ] Modified the constructor of Rental so Book and Person are set in it.

### CODE REVIEW

File: classroom.rb [hunk 1]
```diff
@@ -0,0 +1,13 @@
+class Classroom
+  attr_accessor :label
```
Comment (mikethreels) [inline classroom.rb]: - [ ] in order for this relationship to work the student instance variable should be readable.

File: student.rb [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+
+  def initialize(age, _classroom, name = 'Unknown', parent_permission: true)
+    super(name, age, parent_permission)
+    @classroom = nil
+  end
+
+  def play_hook
+    '¯\(ツ)/¯'
+  end
+end
```
Comment (mikethreels) [inline student.rb]: - [ ] in the Student class, there should be a method that serves as the 'belongs-to' side (a student belongs to a classroom). have a look at [the article with examples](https://github.com/microverseinc/curriculum-ruby/blob/main/oop/articles/oop_relationships_by_examples.md#both-ways-relationships)

File: book.rb [hunk 3]
```diff
@@ -0,0 +1,14 @@
+class Book
+  attr_accessor :title, :author
+  attr_reader :rentals
```
Comment (mikethreels) [inline book.rb]: - [ ] for this relationship to work Getters and Setters for @rentals should be present

File: book.rb [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+
+  def initialize(title, author)
+    @title = title
+    @author = author
+    @rentals = []
+  end
+
+  def add_rental(rental)
+    @rentals << rental
+  end
```
Comment (mikethreels) [inline book.rb]: - [ ] good job creating this method, however, it should look a little different. please review [this article](https://github.com/microverseinc/curriculum-ruby/blob/main/oop/articles/oop_relationships_by_examples.md#many-to-many-relationship) for more details.

File: rental.rb [hunk 5]
```diff
@@ -0,0 +1,14 @@
+require './book'
+
+class Rental
+  attr_accessor :date
+  attr_reader :person, :book
```
Comment (mikethreels) [inline rental.rb]: - [ ] for this relationship to work Getters and Setters for @book and @person instance variables should be present

File: person.rb [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+    @parent_permission || of_age?(@age)
+  end
+
+  def correct_name
+    @name
+  end
+
+  def add_rental(rental)
+    @rentals << rental
+  end
```
Comment (mikethreels) [inline person.rb]: - [ ] good job creating this method, however, it should look a little different. please review [this article](https://github.com/microverseinc/curriculum-ruby/blob/main/oop/articles/oop_relationships_by_examples.md#many-to-many-relationship) for more details.

**Review by mikethreels — CHANGES_REQUESTED** [review]
Hi Evans, 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

## Highlights
- good job starting with the associations.

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: classroom.rb (hunk 1 above)
Comment (Munsa1) [reply classroom.rb]: - [ ] As requested kindly add the student instance variable and make it readable.

e.g

`attr_reader :students`

File: book.rb [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+    @title = title
+    @author = author
+    @rentals = []
+  end
+
+  def add_rental(rental)
+    return if @rentals.include?(rental)
+
+    @rentals << rental
+  end
```
Comment (Munsa1) [inline book.rb]: - [ ] (Optional) Good job adding the add_rental method. Your approach assumes that rentals are created elsewhere and focuses on managing a collection of existing Rental objects. Kindly consider using the approach I have proposed below as it is a little more direct.

e.g
```ruby
def add_rental(person, date) 
    Rental.new(date, self, person)
  end
```
 .

File: person.rb [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+    @parent_permission || of_age?(@age)
+  end
+
+  def correct_name
+    @name
+  end
+
+  def add_rental(rental)
+    @rentals << rental unless @rentals.include?(rental)
+  end
```
Comment (Munsa1) [inline person.rb]: - [ ]  (Optional) Good job implementing this method as requested however I would advise you to implement as follows as this is more direct. Your approach assumes that rentals are created elsewhere and focuses on managing a collection of existing Rental objects.

e.g

```ruby
def add_rental(book, date)
    Rental.new(date, book, self)
  end
```

**Review by Munsa1 — CHANGES_REQUESTED** [review]
Hi @prg-04, 

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.


 ## Highlights
Progress Made :heavy_check_mark: 

 ## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by Reem-lab — APPROVED** [review]
# Hi @prg-04,  👋🏻👋🏻

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 


## Highlights

- [ ]  Followed Gitflow 💯
- [ ]  Descriptive PR⭐
- [ ]  Kudos for working on the associations🤩
- [ ]  No Linters error ✅
- [ ] Good job adding the methods for the classes 🙏🏻💯


## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**